R/write_zarr_helpers.R

```diff
@@ -0,0 +1,553 @@
+
+  # Add encoding
+  write_zarr_encoding(store, name, type, version)
+
+  # Write shape attribute
+  Rarr::write_zarr_attributes(file.path(store, name), list(shape = dim(value)))
+}
+
+#' Write Zarr nullable boolean
+#'
+#' Write a nullable boolean to a Zarr store
+#'
+#' @noRd
+#'
+#' @param value Value to write
+#' @param store A Zarr store instance
+#' @param name Name of the element within the Zarr store
+#' @param compression The compression to use when writing the element. Can be
+#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
+#' @param version Encoding version of the element to write
+write_zarr_nullable_boolean <- function(value, store, name, compression, version = "0.1.0", overwrite = FALSE) {
+  # write mask and values
+  create_zarr_group(store, name)
+  value_no_na <- value
+  value_no_na[is.na(value_no_na)] <- FALSE
+
+  zarr_write_compressed(store, paste0(name, "/values"), value_no_na, compression, overwrite = overwrite)
+  zarr_write_compressed(store, paste0(name, "/mask"), is.na(value), compression, overwrite = overwrite)
+
+  # Write attributes
+  write_zarr_encoding(store, name, "nullable-boolean", version)
+}
+
+#' Write Zarr nullable integer
+#'
+#' Write a nullable integer to a Zarr store
+#'
+#' @noRd
+#'
+#' @param value Value to write
+#' @param store A Zarr store instance
+#' @param name Name of the element within the Zarr store
+#' @param compression The compression to use when writing the element. Can be
+#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
+#' @param version Encoding version of the element to write
+write_zarr_nullable_integer <- function(value, store, name, compression, version = "0.1.0", overwrite = FALSE) {
+  # write mask and values
+  create_zarr_group(store, name)
+  value_no_na <- value
+  value_no_na[is.na(value_no_na)] <- -1L
+
+  zarr_write_compressed(store, paste0(name, "/values"), value_no_na, compression, overwrite = overwrite)
+  zarr_write_compressed(store, paste0(name, "/mask"), is.na(value), compression, overwrite = overwrite)
+
+  # Write attributes
+  write_zarr_encoding(store, name, "nullable-integer", version)
+}
+
+#' Write Zarr string array
+#'
+#' Write a string array to a Zarr store
+#'
+#' @noRd
+#'
+#' @param value Value to write
+#' @param store A Zarr store instance
+#' @param name Name of the element within the Zarr store
+#' @param compression The compression to use when writing the element. Can be
+#' one of `"none"`, `"gzip"` or `"lzf"`. Defaults to `"none"`.
+#' @param version Encoding version of the element to write
+write_zarr_string_array <- function(value,
+                                    store,
+                                    name,
+                                    compression,
+                                    version = "0.2.0",
+                                    overwrite = FALSE) {
+
+  if (!is.null(dim(value))) {
+    dims <- dim(value)
+  } else {
+    dims <- length(value)
+  }
```

Comment:
If you're using a recent version of R:

```suggestion
  dims <- dim(value) %||% length(value)
```

(you can also redefine the operator in this pkg for compatibility with older R versions)

Reply:
oh I will change this directly, anndataR requires 4.5 and later anyways. Thanks Hugo.